tensorpipe/common/cuda.h

```diff
@@ -125,14 +127,66 @@ inline int cudaDeviceForPointer(const void* ptr) {
   return attrs.device;
 }
 
-using CudaPinnedBuffer = std::shared_ptr<uint8_t>;
+class CudaPinnedBufferAllocator {
+ public:
+  CudaPinnedBufferAllocator(const CudaPinnedBufferAllocator&) = delete;
+  CudaPinnedBufferAllocator(CudaPinnedBufferAllocator&&) = delete;
+  CudaPinnedBufferAllocator& operator=(const CudaPinnedBufferAllocator&) =
+      delete;
+  CudaPinnedBufferAllocator& operator=(CudaPinnedBufferAllocator&&) = delete;
+
+  CudaPinnedBufferAllocator(size_t numChunks = 1024, size_t chunkSize = 1024)
+      : kNumChunks_(numChunks),
+        kChunkSize_(chunkSize),
+        chunkAvailable_(numChunks, false),
+        basePtr_(allocPinnedBuffer(numChunks * chunkSize)) {}
+
+  std::shared_ptr<uint8_t> getBuffer(size_t size) {
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      size_t requestedChunks = (size + kChunkSize_ - 1) / kChunkSize_;
+      size_t startChunk = 0;
+
+      // Linear check for available contiguous chunks.
+      for (size_t curChunk = 0; curChunk < kNumChunks_; ++curChunk) {
+        if (!chunkAvailable_[curChunk]) {
+          startChunk = curChunk + 1;
+        } else if (curChunk - startChunk + 1 == requestedChunks) {
+          for (size_t i = startChunk; i <= curChunk; ++i) {
+            chunkAvailable_[i] = false;
+          }
+
+          return std::shared_ptr<uint8_t>(
+              basePtr_.get() + startChunk * kChunkSize_,
+              [this, startChunk, curChunk](uint8_t* /* unused */) {
+                for (size_t i = startChunk; i <= curChunk; ++i) {
+                  // There is a harmless race condition here.
+                  chunkAvailable_[i] = true;
+                }
+              });
```

Comment:
The assumption here is that the allocator will not be destroyed until all channels in the context have invoked all their callbacks, so every chunk will have been reclaimed by then.